clippy_lints/src/types.rs

```diff
@@ -1215,12 +1215,14 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Casts {
                 if from_align < to_align;
                 // with c_void, we inherently need to trust the user
                 if !is_c_void(cx, from_ptr_ty.ty);
+                // when casting from a ZST, we don't know enough to properly lint
+                if !(Some(true) == cx.layout_of(from_ptr_ty.ty).ok().map(|a| a.is_zst()));
```

Comment:
probably should just have a `let Some(from_layout) = ...` near the top, and then derive `from_align` from it as well as performing the ZST check

Reply:
Give me a second.